sdk/iot/hub/tests/cmocka/test_az_iot_hub_client_twin.c

```diff
@@ -0,0 +1,437 @@
+
+  assert_precondition_checked(az_iot_hub_client_twin_get_publish_topic_get(
+      &client, test_device_request_id, test_span, NULL));
+}
+
+static void test_az_iot_hub_client_twin_patch_publish_topic_get_NULL_client_fails(void** state)
+{
+  (void)state;
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_patch_pub_topic) - 1];
+  az_span test_span = az_span_init(test_span_buffer, 0, _az_COUNTOF(test_span_buffer));
+
+  assert_precondition_checked(az_iot_hub_client_twin_patch_publish_topic_get(
+      NULL, test_device_request_id, test_span, &test_span));
+}
+
+static void test_az_iot_hub_client_twin_patch_publish_topic_get_invalid_request_id_fails(
+    void** state)
+{
+  (void)state;
+
+  az_iot_hub_client client;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, NULL), AZ_OK);
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_patch_pub_topic) - 1];
+  az_span test_span = az_span_init(test_span_buffer, 0, _az_COUNTOF(test_span_buffer));
+
+  uint8_t test_bad_request_id_buf[1];
+  az_span test_bad_request_id
+      = az_span_init(test_bad_request_id_buf, 0, _az_COUNTOF(test_bad_request_id_buf));
+  test_bad_request_id._internal.ptr = NULL;
+
+  assert_precondition_checked(az_iot_hub_client_twin_patch_publish_topic_get(
+      &client, test_bad_request_id, test_span, &test_span));
+}
+
+static void test_az_iot_hub_client_twin_patch_publish_topic_get_NULL_span_fails(void** state)
+{
+  (void)state;
+
+  az_iot_hub_client client;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, NULL), AZ_OK);
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_patch_pub_topic) - 1];
+  az_span test_span = az_span_init(test_span_buffer, 0, _az_COUNTOF(test_span_buffer));
+  test_span._internal.ptr = NULL;
+
+  assert_precondition_checked(az_iot_hub_client_twin_patch_publish_topic_get(
+      &client, test_device_request_id, test_span, &test_span));
+}
+
+static void test_az_iot_hub_client_twin_patch_publish_topic_get_NULL_out_span_fails(void** state)
+{
+  (void)state;
+
+  az_iot_hub_client client;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, NULL), AZ_OK);
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_patch_pub_topic) - 1];
+  az_span test_span = az_span_init(test_span_buffer, 0, _az_COUNTOF(test_span_buffer));
+
+  assert_precondition_checked(az_iot_hub_client_twin_patch_publish_topic_get(
+      &client, test_device_request_id, test_span, NULL));
+}
+
+#endif // NO_PRECONDITION_CHECKING
+
+static void test_az_iot_hub_client_twin_response_subscribe_topic_filter_get_succeed(void** state)
+{
+  (void)state;
+
+  az_iot_hub_client client;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, NULL), AZ_OK);
+
+  uint8_t test_span_buffer[_az_COUNTOF(test_correct_twin_response_topic_filter) - 1];
```

Comment:
I'd actually initialize this with fixed space (128, 256) to:

1. demonstrate the way we need app-developers to think about the buffers.
2. check to see the buffer is not overrun

Reply:
I kinda see your point about demonstrating use case but I'm not so sure that unit tests is the place where people should go to understand how to use API's.
As for the buffer overrun part I can see that too.
I would like to issue another PR that checks the span length for all tests. @hihigupt had brought that up previously and I think it's an sdk test improvement. Is it okay to push to another PR focusing solely on that?